### ISSUE TITLE
ConnectionHandler throws NullPointerException in login() after upgrading from 1.0.14

### ISSUE THREAD

**Issue by Linnun** [TheHolyWaffle/TeamSpeak-3-Java-API#307]
I tried to upgrade from 1.0.14 to the latest build (current master branch, fresh from GitHub).

However, with that version I get the following error upon starting the bot:

```
[main] ERROR com.github.theholywaffle.teamspeak3.TS3Query - ConnectionHandler threw exception in connect handler
java.lang.NullPointerException
        at com.github.theholywaffle.teamspeak3.TS3Query.doCommandAsync(TS3Query.java:226)
        at com.github.theholywaffle.teamspeak3.TS3ApiAsync.executeAndReturnError(TS3ApiAsync.java:5348)
        at com.github.theholywaffle.teamspeak3.TS3ApiAsync.login(TS3ApiAsync.java:3449)
        at com.github.theholywaffle.teamspeak3.TS3Api.login(TS3Api.java:3000)
        at XXX.bot.Bot.onConnect(Bot.java:2111)
```

The lines in Bot:2111 are basically just this:

		api = query.getApi();
		api.login("Bot", "secret"); // 2111

Downgrading back to 1.0.14 of this API solves the problem and it works flawlessly again. Is this a known issue when upgrading? I couldn't find any backwards incompatibility notes.

**Comment by rogermb** [comment]
First, over 2000 lines? O.o
I know TS3Api and TS3ApiAsync are guilty of this, too, but I can't really change that original design without breaking everyone's code.

You're almost certainly using the wrong `TS3Query` or `TS3Api` in your onConnect handler. Could you maybe show me the first few lines of your `onConnect` method?

Also, could you please clarify what you mean by "the latest build"? Did you download v1.2.0 or did you build the project yourself?

**Comment by Linnun** [comment]
Yeah, more than 2000 lines in the main file of the bot. It has quite some work to do ;)

This is the entire onConnect method you requested:

```
	@Override
	public void onConnect(TS3Query ts3Query) {
		api = query.getApi();
		if (DEBUG) {
			api.login("BotBeta", "secret");
		} else {
			api.login("Bot", "secret"); // line 2111
		}
		api.selectVirtualServerById(1);
		api.setNickname(botname);
		final int clientId = whoAmI().getId();
		api.moveClient(clientId, botChannel);
		api.registerAllEvents();

		syncChannelsWithDatabase();
	}
```


Prior to that, the following lines setup the ts-api-related variables

```
	// TS3 ServerQuery API
	private final TS3Config config = new TS3Config();

	{
		if (DEBUG) {
			config.setHost("secret host");
			config.setQueryPort(10011);
		} else {
			config.setFloodRate(FloodRate.UNLIMITED);
			config.setHost("127.0.0.1");
			config.setQueryPort(10011);
		}
		config.setReconnectStrategy(ReconnectStrategy.exponentialBackoff());
		config.setConnectionHandler(this);
	}

	private final TS3Query query;

	{
		query = new TS3Query(config);
		query.connect();

		try {
			Thread.sleep(10000);	// I added this part ages ago and don't remember exactly why. I think I once had issues with the bot starting to do bot-stuff before the ts connection has been fully established, so I added this generous time to establish a proper ts connection.
		} catch (InterruptedException e) {
			log.warning("Thread-sleeping interrupted");
		}
	}

	private TS3Api api;

	{
		api = query.getApi();
		api.addTS3Listeners(this);
	}

	private TS3ApiAsync apiAsync;

	{
		apiAsync = query.getAsyncApi();
	}
```

**Comment by Linnun** [comment]
I have downloaded the jar files of the latest releases, and tested them one by one, just to make sure.

teamspeak3-api-1.0.14.jar works perfectly fine.
teamspeak3-api-1.1.0.jar and teamspeak3-api-1.2.0-with-dependencies.jar result in the same error:

```
java.lang.NullPointerException
        at com.github.theholywaffle.teamspeak3.TS3Query.doCommandAsync(TS3Query.java:226)
        at com.github.theholywaffle.teamspeak3.TS3ApiAsync.executeAndReturnError(TS3ApiAsync.java:5258)
        at com.github.theholywaffle.teamspeak3.TS3ApiAsync.login(TS3ApiAsync.java:3578)
        at com.github.theholywaffle.teamspeak3.TS3Api.login(TS3Api.java:2924)
        at XXX.bot.Bot.onConnect(Bot.java:2111)
        at com.github.theholywaffle.teamspeak3.api.reconnect.ReconnectingConnectionHandler.onConnect(ReconnectingConnectionHandler.java:56)
        at com.github.theholywaffle.teamspeak3.TS3Query.connect(TS3Query.java:131)
        at XXX.bot.Bot.<init>(Bot.java:606)
        at XXX.bot.Bot.main(Bot.java:692)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.eclipse.jdt.internal.jarinjarloader.JarRsrcLoader.main(JarRsrcLoader.java:61)
```

**Comment by rogermb** [comment]
```java
	@Override
	public void onConnect(TS3Query ts3Query /*<-- this is the query we're giving you to connect */) {
		api = query.getApi(); /* <-- this is the query you're using, for some reason */
		if (DEBUG) {
			api.login("BotBeta", "secret");
		} else {
			api.login("Bot", "secret"); // line 2111
		}
		api.selectVirtualServerById(1);
		api.setNickname(botname);
		final int clientId = whoAmI().getId();
		api.moveClient(clientId, botChannel);
		api.registerAllEvents();

		syncChannelsWithDatabase();
	}
```

Basically, here's the deal. If you want to send a command that **skips the command queue**, you need to use the `ts3Query` that we give you as the argument in the `onConnect` method. But outside the `onConnect` method, you **cannot** use that TS3Query.

For that reason, please also delete that line `api = query.getApi();`. Just assign that `api` field **once** after creating the `TS3Query` object.

i.e. try

```java
	@Override
	public void onConnect(TS3Query ts3Query) {
		TS3Api api = ts3Query.getApi();
		if (DEBUG) {
			api.login("BotBeta", "secret");
		} else {
			api.login("Bot", "secret"); // line 2111
		}
		api.selectVirtualServerById(1);
		api.setNickname(botname);
		final int clientId = whoAmI().getId();
		api.moveClient(clientId, botChannel);
		api.registerAllEvents();

		syncChannelsWithDatabase();
	}
```

**Comment by Linnun** [comment]
Hey, that seems to work!

The bot complains about stuff that is being done in syncChannelsWithDatabase() now, but login and everything after being connected works fine.

I assume syncChannelsWithDatabase() (e.g. all sub-methods of `onConnect()`) should also use the ts3Query instance provided in onConnect()? Gonna work on that next, then it might be running smoothly again.

Thank you for your help! :)

**Comment by rogermb** [comment]
> The bot complains about stuff that is being done in syncChannelsWithDatabase() now

That's interesting, what does it complain about? 😄 
You *should be able to* use your regular API object from within `onConnect`, but the commands dispatched with it will only run after the startup commands have run.

**Comment by Linnun** [comment]
It had the same kind of NPE thrown by the ConnectionHandler again, but this time caused by an api.getChannels() call within syncChannelsWithDatabase().

I made all api calls within syncChannelsWithDatabase() use the api instance from onConnect, and now it works just fine :)

**Comment by rogermb** [comment]
Interesting! If you have a few minutes to spare, could you try this again using the [latest snapshot release](https://oss.sonatype.org/content/repositories/snapshots/com/github/theholywaffle/teamspeak3-api/1.3.0-SNAPSHOT/teamspeak3-api-1.3.0-20190123.203943-10-with-dependencies.jar)? I've rewritten parts of the IO system, especially the parts that deal with reconnects.

To get your code to work, you'd have to change
```java
	@Override
	public void onConnect(TS3Query ts3Query) {
		TS3Api api = ts3Query.getApi();
```
to
```java
	@Override
	public void onConnect(TS3Api api) {
```

Basically, I wonder if you still get an NPE if you were to use the queue-skipping-API in the onConnect method but the regular API in syncChannelsWithDatabase()

**Comment by rogermb** [comment]
Oh, actually, I think I just figured out why you're getting an NPE.
You probably have code like this somewhere, right?

```java
query = new TS3Query(config);
query.connect();
api = query.getApi();
```

Now, the `onConnect` method gets executed inside `query.connect()` when you connect, so at that point, you simply haven't set your `api` field yet. So essentially, the fix for this issue is to just set `api` earlier:

```java
query = new TS3Query(config);
api = query.getApi();
query.connect();
```

**Comment by Linnun** [comment]
Interesting theory. I tried that by undoing the changes to syncChannelsWithDatabase() and setting the api variable before connecting, just like you suggested. However, that again results in the NPE.


---

I tried your snapshot release. It actually freezes without any error message upon the first api call inside syncChannelsWithDatabase.

I added a debug output line between every single line, and this is basically how far it goes:

```
SNAPSHOT-DEBUG: setting up TS3Api
SNAPSHOT-DEBUG: adding listener
SNAPSHOT-DEBUG: calling connect
SNAPSHOT-DEBUG: onConnect called
SNAPSHOT-DEBUG: calling login now
SNAPSHOT-DEBUG: login done
SNAPSHOT-DEBUG: selecting vserver
SNAPSHOT-DEBUG: setting nickname
SNAPSHOT-DEBUG: calling whoami
SNAPSHOT-DEBUG: moving to botchannel
SNAPSHOT-DEBUG: registering events
SNAPSHOT-DEBUG: syncing channels with db		// this is the print right before calling syncChannelsWithDatabase
SNAPSHOT-DEBUG: syncChannelsWithDatabase called		
SNAPSHOT-DEBUG: calling getChannels
```

This is the body of syncChannelsWithDatabase():

```
	private void syncChannelsWithDatabase() {
		System.out.println("SNAPSHOT-DEBUG: syncChannelsWithDatabase called");
		Date syncDate = new Date();

		// Sync all channels
		System.out.println("SNAPSHOT-DEBUG: calling getChannels");
		List<Channel> channels = api.getChannels();
		System.out.println("SNAPSHOT-DEBUG: got channels: " + channels.size());
		System.out.println("SNAPSHOT-DEBUG: starting foreach channel loop");
		for (Channel channel : channels) {
			syncChannelIntoDatabase(channel.getId());
		}
		System.out.println("SNAPSHOT-DEBUG: ending foreach channel loop");

		// Delete channels that have not been synced
		db.deleteOutdatedChannels(syncDate);
	}
```


So essentially the line `List<Channel> channels = api.getChannels();` causes the entire script to freeze. No further outputs. The bot is on the server, but won't react to messages sent to him.

As you requested, the variable api is now being set before connecting:

```
	private TS3Api api;
	private TS3ApiAsync apiAsync;
	private final TS3Query query;

	{
		System.out.println("SNAPSHOT-DEBUG: setting up TS3Query");
		query = new TS3Query(config);
		System.out.println("SNAPSHOT-DEBUG: setting up TS3Api");
		api = query.getApi();
		apiAsync = query.getAsyncApi();
		System.out.println("SNAPSHOT-DEBUG: adding listener");
		api.addTS3Listeners(this);
		System.out.println("SNAPSHOT-DEBUG: calling connect");
		query.connect();
		System.out.println("SNAPSHOT-DEBUG: connect done");

		try {
			Thread.sleep(10000);
		} catch (InterruptedException e) {
			log.warning("Thread-sleeping interrupted");
		}
	}
```

**Comment by rogermb** [comment]
Oh, right. That **can't** work. I'm feeling so stupid right now 😂 
The query can't continue sending commands from the global command queue because your `onConnect` method hasn't returned yet, but your `onConnect` method can't make any progress until the command that you've enqueued into the global command queue has finished. It's a deadlock!

**Comment by rogermb** [comment]
So yeah, I don't think there's a good way to enforce this programmatically, but it boils down to this:
- Use the provided TS3Api and don't use your query's TS3Api while in onConnect
- Use your query's TS3Api and don't use an onConnect-TS3Api outside an onConnect method

¯\\\_(ツ)_/¯

Thanks for the testing! 😄

**Comment by Linnun** [comment]
No worries. Now that we tested a few things, we can see what happened and why certain things didn't work. Please let me know if I can assist with any additional testing :)
